### PR TITLE
tests: Support multiple keys in linearizability tests

### PR DESCRIPTION
**Pull request by serathius**
cc @ahrtr @spzala @tjungblu

### CODE REVIEW

File: tests/linearizability/model.go [hunk 1]
```diff
@@ -47,8 +47,8 @@ type EtcdResponse struct {
 }
 
 type EtcdState struct {
-	Key            string
-	PossibleValues []ValueRevision
+	Revision  int64
+	KeyValues map[string]string
 }
 
 type ValueRevision struct {
```
Comment (ahrtr) [inline tests/linearizability/model.go:54]: Why do you still keep `ValueRevision`?

File: tests/linearizability/model.go [hunk 2]
```diff
@@ -105,92 +105,99 @@ var etcdModel = porcupine.Model{
...
 		panic("Unknown operation")
 	}
 }
 
-func stepValue(v ValueRevision, request EtcdRequest) (ValueRevision, EtcdResponse) {
+func stepState(s EtcdState, request EtcdRequest) (EtcdState, EtcdResponse) {
+	newKVs := map[string]string{}
+	for k, v := range s.KeyValues {
+		newKVs[k] = v
+	}
```
Comment (ahrtr) [inline tests/linearizability/model.go:177]: Please consider to add a clone method something like below,

```
func (s EtcdState) clone() EtcdState {
	newKVs := map[string]string{}
	for k, v := range s.KeyValues {
		newKVs[k] = v
	}  
	return EtcdState{
		Revision: s.Revision,
		KeyValues: newKVs,
	}
}

```

File: tests/linearizability/model.go [hunk 3]
```diff
@@ -47,8 +47,8 @@ type EtcdResponse struct {
 }
 
 type EtcdState struct {
-	Key            string
-	PossibleValues []ValueRevision
+	Revision  int64
```
Comment (ahrtr) [inline tests/linearizability/model.go:50]: Based on the implementation in this PR, it seems no need to keep `Revision` in `EtcdState`?

File: tests/linearizability/model.go [hunk 4]
```diff
@@ -47,8 +47,8 @@ type EtcdResponse struct {
...
+	KeyValues map[string]string
```
Comment (ahrtr) [inline tests/linearizability/model.go:51]:
```suggestion
	KeyValues map[string][]ValueRevision
```

File: tests/linearizability/model.go (hunk 1 above)
Comment (ahrtr) [reply tests/linearizability/model.go:54]: Based on https://github.com/etcd-io/etcd/pull/14924#discussion_r1045433830, we still need to keep `ValueRevision`

File: tests/linearizability/model.go (hunk 2 above)
Comment (ahrtr) [reply tests/linearizability/model.go:177]: This comment isn't valid any more given https://github.com/etcd-io/etcd/pull/14924/files#r1045433830

File: tests/linearizability/model.go [hunk 5]
```diff
@@ -57,15 +57,15 @@ type ValueRevision struct {
 }
 
 var etcdModel = porcupine.Model{
-	Init: func() interface{} { return "{}" },
+	Init: func() interface{} { return "[]" },
```
Comment (ahrtr) [inline tests/linearizability/model.go:60]:
```suggestion
	Init: func() interface{} { return "{}" },
```

File: tests/linearizability/model.go [hunk 6]
```diff
@@ -57,15 +57,15 @@ type ValueRevision struct {
...
 	Step: func(st interface{}, in interface{}, out interface{}) (bool, interface{}) {
-		var state EtcdState
-		err := json.Unmarshal([]byte(st.(string)), &state)
+		var states []EtcdState
```
Comment (ahrtr) [inline tests/linearizability/model.go:62]:
```suggestion
		var state EtcdState
```

File: tests/linearizability/model.go [hunk 7]
```diff
@@ -105,92 +105,99 @@ var etcdModel = porcupine.Model{
 	},
 }
 
-func step(state EtcdState, request EtcdRequest, response EtcdResponse) (bool, EtcdState) {
-	if len(state.PossibleValues) == 0 {
-		state.Key = request.Key
-		if ok, v := initValueRevision(request, response); ok {
-			state.PossibleValues = append(state.PossibleValues, v)
+func step(states []EtcdState, request EtcdRequest, response EtcdResponse) (bool, []EtcdState) {
```
Comment (ahrtr) [inline tests/linearizability/model.go:108]:
```suggestion
func step(state EtcdState, request EtcdRequest, response EtcdResponse) (bool, EtcdState) {
```

File: tests/linearizability/model.go [hunk 8]
```diff
@@ -105,92 +105,99 @@ var etcdModel = porcupine.Model{
...
+	if len(states) == 0 {
```
Comment (ahrtr) [inline tests/linearizability/model.go:109]:
```suggestion
	if _, ok := state.KeyValues[request.Key]; !ok {
```

File: tests/linearizability/model.go [hunk 9]
```diff
@@ -105,92 +105,99 @@ var etcdModel = porcupine.Model{
...
-	}
-	if state.Key != request.Key {
-		panic("multiple keys not supported")
+		return true, states
 	}
 	if response.Err != nil {
-		for _, v := range state.PossibleValues {
-			newV, _ := stepValue(v, request)
-			state.PossibleValues = append(state.PossibleValues, newV)
+		for _, s := range states {
```
Comment (ahrtr) [inline tests/linearizability/model.go:116]:
```suggestion
		kvs, _ := state.KeyValues[request.Key]
		for _, v := range kvs {		
```

File: tests/linearizability/model.go [hunk 10]
```diff
@@ -105,92 +105,99 @@ var etcdModel = porcupine.Model{
...
+			newState, _ := stepState(s, request)
+			states = append(states, newState)
 		}
 	} else {
 		var i = 0
-		for _, v := range state.PossibleValues {
-			newV, expectedResponse := stepValue(v, request)
+		for _, s := range states {
```
Comment (ahrtr) [inline tests/linearizability/model.go:122]: similar to https://github.com/etcd-io/etcd/pull/14924/files#r1045440291

File: tests/linearizability/model.go (hunk 5 above)
Comment (serathius) [reply tests/linearizability/model.go:60]: I changed storing `etcdState` to storing `[]etcdState`. Are you saying it's a bad change or don't understand the purpose?

**Comment by serathius** [comment]
Resolved all comments that do not make sense. Please describe what you would like to change and not send 20 suggestions that will not work.

Suggestions make sense only if:
* They don't affect rest of code
* Change all code that is affected at once https://github.blog/changelog/2020-02-26-multi-line-code-suggestions-beta/

File: tests/linearizability/model.go (hunk 1 above)
Comment (serathius) [reply tests/linearizability/model.go:54]: Removed.

File: tests/linearizability/model.go [hunk 11]
```diff
@@ -105,92 +100,99 @@ var etcdModel = porcupine.Model{
...
+			newState, expectedResponse := stepState(s, request)
 			if expectedResponse == response {
-				state.PossibleValues[i] = newV
+				states[i] = newState
 				i++
 			}
 		}
-		state.PossibleValues = state.PossibleValues[:i]
+		states = states[:i]
```
Comment (tjungblu) [inline tests/linearizability/model.go:124]: clever idea with the state reduction.

**Review by tjungblu — COMMENTED** [review]
lgtm (non-binding)

File: tests/linearizability/model.go (hunk 11 above)
Comment (serathius) [reply tests/linearizability/model.go:124]: Yea, but looks like it's not very readable. Rewrote it to make the code cleaner.

**Comment by ahrtr** [comment]
Overall looks good to me. 

The model is the most important part of the linearizablity test, so please add detailed comment to describe the overall thought, especially you are changing the code too fast and frequently. I should raise this comment in your last PR in which you totally changed the implementation. Otherwise others may spend lots of time to understand your thoughts.

**Comment by serathius** [comment]
Decided to separate PR that refactors the model first. Let's discuss documenting model there. https://github.com/etcd-io/etcd/pull/15045

**Comment by ahrtr** [comment]
Please add self-contained comment in `model.go` to describe the overall thought.  And also add comment something like below for `stepState`,
```
`stepState` generates all the possible expected states and expected response based on all the current states. 
The caller can compare the expected response with the actual response, and take action accordingly.
```

**Comment by serathius** [comment]
@ahrtr Done, please take a look.

File: tests/linearizability/model.go [hunk 12]
```diff
@@ -101,87 +104,90 @@ var etcdModel = porcupine.Model{
...
 		}
-		return []EtcdState{}
 	default:
 		panic("Unknown operation")
 	}
-	return []EtcdState{state}
+	return state
+}
+
+// applyFailedRequest handles a failed requests, one that it's not known if it was persisted nor not.
```
Comment (ahrtr) [inline tests/linearizability/model.go:147]:
```suggestion
// applyFailedRequest handles a failed requests, one that it's not known if it was persisted or not.
```

**Review by ahrtr — APPROVED** [review]

**Pull request merged**